### PR TITLE
New doc: Working with Cookies

### LINKED ISSUE #414: Document how to send HTTP headers that Varnish will care about
Story: As a website developers I want to understand how to send HTTP headers that Varnish will care about so that I can control my websites cache behavior.

Right now we have people setting cookies as a way to prevent pages from being cached, which is bass-ackwards. We should start teaching people about HTTP headers and how to set them to control downstream caching.

Problem: I want to control cacheability and/or cache-lifetime on a per-page basis
Soution: get smart about how to use the internal page cache, and HTTP headers for downstream (varnish) caching
About HTTP headers
How/why to use both the internal page cache and the downstream page cache
Customizing cache lifetime
Preventing a single page from caching

--submitted by @joshkoenig

### LINKED ISSUE #1581: New Doc Request: Working with Cookies on Pantheon
Currently we we bury a lot of pretty key info at the bottom of a section here:

https://pantheon.io/docs/debug-cache/#cookies-&-varnish

For many web developers, "Varnish" is a foreign word, or at least some advanced thing they don't understand. We should consider creating a doc that's focused around how to do things with websites and cookies, laying out the important basic info above, and the specific cookie tips and tricks from this page:

https://pantheon.io/docs/caching-advanced-topics/

### PR DESCRIPTION
**Pull request by nataliejeremy**
Closes #1581 
Closes #414 

## Effects 
- New Doc: **Bypassing Cache with HTTP Headers**
- New Doc: **Working with Cookies on Pantheon**
- Refactored **Understanding and Debugging Varnish Cache Issues**, renamed **Clearing Caches for Drupal and WordPress** (requires redirect once deployed)
- Changed title from **Varnish Caching - Drupal and WordPress Advanced Topics** to **Caching: Advanced Topics** (no redirect required)
- Moved troubleshooting section to **Varnish Caching for High Performance with Drupal and WordPress**

## TODO

- [x] Simplify intro to Varnish and use "Pantheon's edge cache" within supporting content re: cookies
- [x] Build examples for D7/D8/WP demonstrating how to bypass cache using cookies
  - Include example scenarios where different methods would be used
- [x] Audit existing content on cookies to reduce duplicate content
  - [Understanding and Debugging Varnish Cache Issues](https://pantheon.io/docs/debug-cache/#cookies-&-varnish) 
  - [Varnish Caching - Drupal and WordPress Advanced Topics](https://pantheon.io/docs/caching-advanced-topics/)

### CODE REVIEW

**Comment by rachelwhitton** [comment]
```
// Cache-varying cookie single page
// Replace /page with desired path
if ($_SERVER['REQUEST_URI'] == '/page') {
  if(isset($_COOKIE['STYXKEYcookie'])){
      $cookie = $_COOKIE['STYXKEYcookie'];
  }
  else{
    setcookie("STYXKEYcookie", "test", time()+3600*1, "/page", $_SERVER['HTTP_HOST']);
  }
}
```

**Comment by nataliejeremy** [comment]
@allenfear Who would you recommend to review this PR?

**Comment by nataliejeremy** [comment]
@ttrowell Can you review and +1 this PR please?

File: source/_docs/cookies.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+If the value has already been set, do not set the cookie again in the response. Varnish cannot cache a response that contains a <code>Set-Cookie:</code> header.
+</div>
+
+If the value is **not** set, respond with `setcookie()` to serve cached content for subsequent requests within the defined cookie lifetime.
+
+The following example can be used interchangeably between WordPress and Drupal:
+```
+if(isset($_COOKIE['STYXKEY_gorp'])){
+
+  $foo = $_COOKIE['STYXKEY_gorp'];
```
Comment (dustinleblanc) [inline source/_docs/cookies.md]: I don't fully understand this code block, why do we assign the value of the cookie? I don't immediately see how it gets used later.

File: source/_docs/mobile-redirection.md [hunk 2]
```diff
@@ -45,7 +45,7 @@ In order for CloudFlare's mobile redirection service to work, the domain must be
 
 
 ### JavaScript
-You can detect the page width or user agent via JavaScript, then redirect requests to your configured and optimized mobile site. We do not recommend using cookies that are passed to the backend for mobile theme detection and configuration. For details, see [Varnish Caching - Drupal and WordPress Advanced Topics](/docs/caching-advanced-topics/#device-detection).
+You can detect the page width or user agent via JavaScript, then redirect requests to your configured and optimized mobile site. We do not recommend using cookies that are passed to the backend for mobile theme detection and configuration. For details, see [Caching: Advanced Topics](/docs/caching-advanced-topics/#device-detection).
```
Comment (dustinleblanc) [inline source/_docs/mobile-redirection.md]: As a developer I cringed anytime someone would check page width with JS. There are tons of known issues with this. Page widths don't calculate reliably with JS. If they want to do page width settings, they should be doing responsive design and doing it completely in css. Checking for user agent seems different though

File: source/_docs/varnish.md [hunk 3]
```diff
@@ -21,10 +21,85 @@ No module or plugin installation is required; do <strong>not</strong> install th
...
 
-For detailed instructions on how to optimize your caching configuration, see [Drupal's Performance Settings](/docs/drupal-cache) or [WordPress Pantheon Cache Plugin Configuration](/docs/wordpress-cache-plugin/).
+For detailed instructions on how to configure and optimize caching, see [Drupal's Performance Settings](/docs/drupal-cache) or [WordPress Pantheon Cache Plugin Configuration](/docs/wordpress-cache-plugin/).
 
+## Troubleshooting Varnish Hits
+### No HTTP Cache Headers in Drupal
+If you have checked your HTTP headers and found that the cache is not working, make sure you have configured [Drupal's performance settings](/docs/drupal-cache). Once you have completed this step, go back and check the HTTP headers to verify that Varnish is working.
+
+If you are still getting `no-cache, must-revalidate, post-check=0, pre-check=0` as a response, check to see if any messages are being set - [drupal\_set\_message](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/drupal_set_message/7) disables page caching. Also check the theme to see if Drupal messages are being set in an attempt to suppress user facing messages.
```
Comment (dustinleblanc) [inline source/_docs/varnish.md]: >Also check the theme to see if Drupal messages are being set in an attempt to suppress user facing messages.

I think we mean "check to see if messages are present in page templates; themers could have removed them to suppress user facing messages."

**Comment by dustinleblanc** [comment]
👍

**Pull request merged**